File: apps/api/plane/app/serializers/phase.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

from rest_framework import serializers

from plane.db.models import Phase, PhaseCycle

from .base import BaseSerializer, DynamicBaseSerializer


class PhaseSerializer(DynamicBaseSerializer):
    """Read serializer — includes annotated cycle counts."""

    total_cycles = serializers.IntegerField(read_only=True, default=0)
    completed_cycles = serializers.IntegerField(read_only=True, default=0)

    class Meta:
        model = Phase
        fields = "__all__"
        read_only_fields = ["workspace", "project", "created_by", "updated_by", "deleted_at"]


class PhaseWriteSerializer(BaseSerializer):
    class Meta:
        model = Phase
        fields = [
            "id",
            "name",
            "description",
            "start_date",
            "end_date",
            "sort_order",
            "archived_at",
        ]


class PhaseCycleSerializer(BaseSerializer):
    class Meta:
        model = PhaseCycle
        fields = "__all__"
        read_only_fields = ["workspace", "project", "created_by", "updated_by", "deleted_at"]

File: packages/types/src/phase.ts
// Copyright (c) 2023-present Plane Software, Inc. and contributors
// SPDX-License-Identifier: AGPL-3.0-only

export type TPhaseStatus = "backlog" | "planned" | "in-progress" | "paused" | "completed" | "cancelled";

export interface IPhase {
  id: string;
  name: string;
  description?: string | null;
  status: TPhaseStatus;
  start_date?: string | null;
  end_date?: string | null;
  lead_id?: string | null;
  member_ids: string[];
  sort_order: number;
  archived_at?: string | null;
  project: string;
  workspace: string;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
  // Annotated counts and flags from the backend
  total_cycles: number;
  completed_cycles: number;
  is_favorite: boolean;
}

export interface IPhaseCreate {
  name: string;
  description?: string;
  status?: TPhaseStatus;
  start_date?: string | null;
  end_date?: string | null;
  lead_id?: string | null;
  member_ids?: string[];
}

export interface IPhaseUpdate extends Partial<IPhaseCreate> {
  sort_order?: number;
  archived_at?: string | null;
}

export interface IPhaseCycle {
  id: string;
  phase: string;
  cycle: string;
  project: string;
  workspace: string;
}
